### PR TITLE
GH-2557: RDF term normalization

### LINKED ISSUE #2557: Provide a framework for normalizing RDF terms
### Version

5.0.0

### Feature

Jena already has support for a fixed canonicalization of literals in `CanonicalizeLiteral`. There is also other code that decides lexical form within the SPARQL expression evaluator.

It would be better called these "normalizations" because there is one fixed choice.

This task is to:

1. Generalize the mechanism to allow different normalization choices e.g. XSD 1.0 vs XSD 1.1; Turtle syntax short-form for numbers; a defined system choice.
2. Provide one form where the canonicalization is preserved when written to a TDB2 database. TDB2 holds binary values for some datatypes and rebuilds the lexical form on retrieval. This effective normalizes. It would be beneficial to provide this as a normalization choice for data and also to include it in the test suite to pin it down.

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #2557

Pull request Description:

Normalization is mostly involved with literals where the lexical form has multiple ways to write the same value. Numbers are the major case: "1", "001", "+1" are ways to write the xsd:integer value 1.

There is now a framework for normalization, with 3 implementations, ARQ choices, XSD 1.0 and XSD 1.1. The XSD differ in a few places: the canonical form of xsd:decimals for integer values, XSD 1.1 adds `+INF`.

The ARQ choices are, as near as possible, the same as previous versions and align with the effect of adding terms to TDB2 and reading them back again. This was the case up to Jena 5.0.0 but wasn't tested. Now it is contract so data can be normalized in a way thatTDB2 will preserve the exact RDF terms for float and double precision.

Many files are changed because of a rename of `NodeFactory.createLiteral(string, datatype) ` to `NodeFactory.createLiteralDT(String, datatype)` to align with other 5.0.0 renaming. A deprecated placeholder is left behind.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+/**
+ * Convert literals to normalized forms. Sometimes called canonicalization. There is
+ * one preferred RDFTerm for a given RDFterm "value" ("value" generalized to include
+ * URis and blank nodes). Mostly, this affects literals. Only certain datatypes are
+ * supported but applications can add normalization for other datatypes.
+ * <p>
+ * Various policies are provided:
+ * <ul>
+ * <li>{@link #get() General} (close to Turtle, use the turtle form for long term contract</li>
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java:40]:
```suggestion
 * <li>{@link #get() General} (close to Turtle, use the turtle form for long term contract)</li>
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+ * <li>{@link #getTTL() Turtle}</li>
+ *  * <li>{@link #getXSD() XSD}</li>
+ * <li>XSD - follows XSD 1.1. Mantissa/exponents are adjusted. xsd;decimal of an
+ * integer value does not have a decimal point. and is not suitable for Turtle as a
+ * xsd;decimal short form.</li>
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java:45]:
```suggestion
 * xsd:decimal short form.</li>
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+ * URis and blank nodes). Mostly, this affects literals. Only certain datatypes are
+ * supported but applications can add normalization for other datatypes.
+ * <p>
+ * Various policies are provided:
+ * <ul>
+ * <li>{@link #get() General} (close to Turtle, use the turtle form for long term contract</li>
+ * <li>{@link #getTTL() Turtle}</li>
+ *  * <li>{@link #getXSD() XSD}</li>
+ * <li>XSD - follows XSD 1.1. Mantissa/exponents are adjusted. xsd;decimal of an
+ * integer value does not have a decimal point. and is not suitable for Turtle as a
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java:44]:
```suggestion
 * integer value does not have a decimal point, and is not suitable for Turtle as a
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+ * one preferred RDFTerm for a given RDFterm "value" ("value" generalized to include
+ * URis and blank nodes). Mostly, this affects literals. Only certain datatypes are
+ * supported but applications can add normalization for other datatypes.
+ * <p>
+ * Various policies are provided:
+ * <ul>
+ * <li>{@link #get() General} (close to Turtle, use the turtle form for long term contract</li>
+ * <li>{@link #getTTL() Turtle}</li>
+ *  * <li>{@link #getXSD() XSD}</li>
+ * <li>XSD - follows XSD 1.1. Mantissa/exponents are adjusted. xsd;decimal of an
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java:43]:
```suggestion
 * <li>XSD - follows XSD 1.1. Mantissa/exponents are adjusted. xsd:decimal of an
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        map.put(XSDDatatype.XSDdecimal,     NormalizeValue.dtDecimalXSD ) ;
+        map.put(XSDDatatype.XSDfloat,       NormalizeValue.dtFloatXSD ) ;
+        map.put(XSDDatatype.XSDdouble,      NormalizeValue.dtDoubleXSD ) ;
+        return map;
+    }
+
+
+    /*
+     * Add the standard set of datatype handlers.
+     * This is the general a policy.
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeRDFTerms.java:230]:
```suggestion
     * This is the general policy.
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java [hunk 6]
```diff
@@ -128,80 +128,112 @@ class NormalizeValue
...
-     * exponent must be indicated by "E". Leading zeroes and the preceding
-     * optional "+" sign are prohibited in the exponent. If the exponent is
-     * zero, it must be indicated by "E0". For the mantissa, the preceding
-     * optional "+" sign is prohibited and the decimal point is required.
-     * Leading and trailing zeroes are prohibited subject to the following:
-     * number representations must be normalized such that there is a single
-     * digit which is non-zero to the left of the decimal point and at least a
-     * single digit to the right of the decimal point unless the value being
-     * represented is zero. The canonical representation for zero is 0.0E0.
+     * Format floats and double by using {@linkDecimalFormat}.
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java:172]:
```suggestion
     * Format floats and double by using {@link DecimalFormat}.
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java [hunk 7]
```diff
@@ -128,80 +128,112 @@ class NormalizeValue
...
+     * This can move the decimal point and change the exponent value.
+     * All numbers are "n.nnnExxx".
+     * For "smaller"floats and double, Java formating as used by
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java:175]:
```suggestion
     * For "smaller" floats and double, Java formatting as used by
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java [hunk 8]
```diff
@@ -128,80 +128,112 @@ class NormalizeValue
...
+     * {@link XSDNumUtils#stringForm(double)} or {@link XSDNumUtils#stringForm(float)}
+     * leaves the number is "common" form, with the mantissa (significand) having the decimal point
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValue.java:177]:
```suggestion
     * leaves the number in "common" form, with the mantissa (significand) having the decimal point
```

File: jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValueByCode.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+import static org.apache.jena.atlas.lib.Chars.CH_DOT;
+import static org.apache.jena.atlas.lib.Chars.CH_MINUS;
+import static org.apache.jena.atlas.lib.Chars.CH_PLUS;
+import org.apache.jena.datatypes.RDFDatatype;
+import org.apache.jena.graph.Node;
+import org.apache.jena.graph.NodeFactory;
+
+/**
+ * Code to format integers an decimals by manipulating the string form
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/riot/process/normalize/NormalizeValueByCode.java:29]:
```suggestion
 * Code to format integers and decimals by manipulating the string form
```

File: jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+ */
+
+package org.apache.jena.sparql.util;
+
+import java.math.BigDecimal;
+
+public class XSDNumUtils {
+
+    /**
+     * Parser an XSD double lexical form.
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java:26]:
```suggestion
     * Parse an XSD double lexical form.
```

File: jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            // Acceptable as Java doubles (value is "NaN" but not as xsd:double
+            case "-NaN"-> throw new NumberFormatException("-NaN is not valid as an xsd:double");
+            case "+NaN"-> throw new NumberFormatException("+NaN is not valid as an xsd:double");
+            // Includes +0 and -0.
+            default-> Double.parseDouble(lexicalForm);
+        };
+    }
+
+    /**
+     * Parser an XSD float lexical form.
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java:47]:
```suggestion
     * Parse an XSD float lexical form.
```

File: jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            return "NaN" ;
+
+        // No SPARQL short form
+        String x = Float.toString(f) ;
+        return x;
+    }
+
+    /**
+     * The format of {@code xsd:decimal} used in ARQ expression evaluation. This is
+     * XSD 1.0 for long-term consistency (integer values for {@code xsd;decimal} have
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java:115]:
```suggestion
     * XSD 1.0 for long-term consistency (integer values for {@code xsd:decimal} have
```

File: jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+        if ( bd.scale() <= 0 )
+            // No decimal part.
+            return bd.toPlainString();
+        return bd.stripTrailingZeros().toPlainString();
+    }
+
+    /**
+     * Integer-valued decimals have a trailing ".0".
+     * (In XML Schema Datatype 1.1 they did not have a ".0".)
+     * <p>
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java:153]:
```suggestion
     *
```

File: jena-tdb2/src/main/java/org/apache/jena/tdb2/sys/NormalizeTermsTDB2.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.jena.riot.process.StreamRDFApply;
+import org.apache.jena.riot.process.normalize.NormalizeRDFTerms;
+import org.apache.jena.riot.system.StreamRDF;
+import org.apache.jena.tdb2.store.NodeId;
+import org.apache.jena.tdb2.store.NodeIdInline;
+
+/**
+ * A variation on {@link NormalizeRDFTerms} that attempts to generate TDB2 {@link NodeIdInline inline} NodeIds,
+ * then convert back to {@link Node Nodes}.
+ * This is used to ensure that TDB2 NodInlining is compatible with {@link NormalizeRDFTerms}
```
Comment (rvesse) [inline jena-tdb2/src/main/java/org/apache/jena/tdb2/sys/NormalizeTermsTDB2.java:31]:
```suggestion
 * This is used to ensure that TDB2 Node Inlining is compatible with {@link NormalizeRDFTerms}
```

**Review by rvesse — APPROVED** [review]
This looks really good

All my comments are just minor typos/grammar issues in the comments, actual implementation code all looks fine as far as I understand the relevant XSD specifications

**Comment by afs** [comment]
> This looks really good
> 
> All my comments are just mfg=typos/grammar issues in the comments, actual implementation code all looks fine as far as I understand the relevant XSD specifications

Yes - this is spec heavy!

File: jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java (hunk 13 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/util/XSDNumUtils.java:153]: The whole line can be removed. Done.

**Comment by afs** [comment]
Corrections merged - thanks.

**Pull request merged**